README.md
A sdk tool for developers use functions of aliyun OCR text recognition 
in the dart language for flutter framework.

## Features
This package supports users skip the process of generating the signature to use
the aliyun OCR text recognition api with HTTP by uploading the image's url in web.

This package supports old and new version of aliyun OCR Method, the old one is Ok
but the new one only support using the url format image but not the binary image.

> [aliyun_ocr_homepage](https://help.aliyun.com/product/252763.html)
> [aliyun signature page](https://help.aliyun.com/document_detail/294534.html)

### Old Method
It's Ok
### New Method
**NOTES:**  In some reason, I can not upload my image with the HTTP ERROR 503.
            After consulting the engineers from aliyun,they tell me to debug
            other framework sdk such as JAVA to find some way and refuse to
            show me how to put my image binary file into the http body.

## Getting started

This package needs user have the aliyun **accessKeyId** and **accessKeySecret**.
Your should login in the website of the aliyun and follow this approach:

> User's image 

> AccessKey Management 

## Usage

```dart

    /// 1.1 Input Your appcode
    aliYunSdkOld.appCode = "YourappCode";
    aliYunSdkOld.generateHeaders();
    
    print(aliYunSdkOld.headers);
    
    /// 1.2 Input ImgUrl
    aliYunSdkOld.imgUrl = "https://img.alicdn.com/tfs/TB1Wo7eXAvoK1RjSZFDXXXY3pXa-2512-3509.jpg";
    aliYunSdkOld.generateBodyWithImg();
    
    /// 1.2 Or Input Img binary
    final value = File("bin/1.png");
    var bytes =await value.readAsBytes();
    aliYunSdkOld.imgBinary = bytes.buffer.asUint8List();
    aliYunSdkOld.generateBodyWithImg();

    /// 2. New Method OCR
    var aliYunSdkNew = aliYunOcrClientNew();

    /// 2.1 Input Your accessKey
    aliYunSdkNew.accessKeyId = "YouraccessKeyId";
    aliYunSdkNew.accessKeySecret = "YouraccessKeySecret";

    /// 2.2 Input imgUrl
    aliYunSdkNew.imgUrl =
    "https://img.alicdn.com/tfs/TB1Wo7eXAvoK1RjSZFDXXXY3pXa-2512-3509.jpg";

    aliYunSdkNew.queries['AccessKeyId'] = aliYunSdkNew.accessKeyId;
    aliYunSdkNew.queries['Url'] = aliYunSdkNew.imgUrl;

    /// 2.3 Generate Signature and update the queries
    aliYunSdkNew.generateSignature(
    accessKeySecret: aliYunSdkNew.accessKeySecret,
    queries: aliYunSdkNew.queries,
    method: "POST",
    );
```

## Additional information

Some issues can be submitted into [My Homepage](https://github.com/legendsonldh/aliyun_ocr_sdk)

This package has been published in pub.dev website, here is the link [aliyun_ocr_sdk 1.0.0](https://pub.dev/packages/aliyun_ocr_sdk)

CHANGELOG.md
## 1.1.0

- Add the old version of aliyun OCR Method

## 1.0.0

- Initial version.


test/aliyun_ocr_sdk_test.dart
import 'package:aliyun_ocr_sdk/aliyun_ocr_sdk.dart';
import 'package:aliyun_ocr_sdk/src/aliyun_ocr_sdk_base.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {

    var aliYunSdk = aliYunOcrClientNew();

    setUp(() {
      // Additional setup goes here.
    });

    test('First Test', () {
    });

  });
}


lib/src/aliyun_ocr_sdk_base.dart
// TODO: Put public facing types in this file.
import 'package:intl/intl.dart';
import 'package:uuid/uuid.dart';
import 'dart:convert';
import 'package:crypto/crypto.dart';

/// The Old Method of aliyunOcr
class aliYunOcrClientOld {
  /// Your Aliyun Account appCode
  String appCode = "";

  /// imgUrl
  String imgUrl = "";

  /// img binary
  var imgBinary;

  /// Http's Headers
  Map<String, String> headers = {};

  /// Method
  Map dict = {};
  
  /// Body
  var body;

  /// aliyun_ocr_serve
  String baseUrl = "";

  /// init the client
  aliYunOcrClientOld() {
    baseUrl = "https://ocrapi-advanced.taobao.com/ocrservice/advanced";

  }

  /// create Headers
  void generateHeaders(){
    headers['Authorization'] = 'APPCODE ' + appCode;
    headers['Content-Type'] = 'application/json; charset=UTF-8';
  }

  /// create body WithUrl
  void generateBodyWithUrl(){
    dict['url'] = imgUrl;
    body = jsonEncode(dict);
  }

  /// create body WithImg
  void generateBodyWithImg(){
    dict['img'] = base64Encode(imgBinary.toList());
    body = jsonEncode(dict);
  }

}

/// The New Method of aliyunOcr
class aliYunOcrClientNew {
  /// Your Aliyun Account accessKey
  String accessKeyId = "";

  /// Your Aliyun Account accessKeySecret
  String accessKeySecret = "";

  /// Use UTC(0) time to generate Signature
  String timeStamp = "";

  /// uuidCode for SignatureNonce
  String uuidCode = "";

  /// HTTP Reponse format "XML" or "JSON"
  String format = "";

  /// OCR Method;  RecognizeGeneral or RecognizeAdvanced
  String action = "";

  /// aliyun_ocr_serve
  String baseUrl = "";

  /// online image link; queries['Url'] = img_url
  String imgUrl = "";

  /// HTTP Method "GET" or "POST"
  String method = "";

  /// Siganture code. It must be generated by algorithm
  String signature = "";

  /// The params your will post or get
  Map<String, String> queries = {
    'Action': 'RecognizeGeneral',
    'Format': 'JSON',
    'AccessKeyId': "",
    'SignatureMethod': 'HMAC-SHA1',
    'Timestamp': "",
    'SignatureVersion': '1.0',
    'Version': '2021-07-07',
    'SignatureNonce': "",
  };

  /// init the Class
  aliYunOcRClientNew() {
    baseUrl = 'http://ocr-api.cn-hangzhou.aliyuncs.com/';
    var now = DateTime.now().toUtc();
    timeStamp =
        '${DateFormat("yyyy-MM-dd").format(now)}T${DateFormat.Hms().format(now)}Z';
    uuidCode = Uuid().v4().toString();
    queries['Timestamp'] = timeStamp;
    queries['SignatureNonce'] = uuidCode;
  }

  /// Signature Algorithm
  void generateSignature({
    required String accessKeySecret,
    required Map<String, String> queries,
    String method = 'GET',
  }) {
    final List<String> _sortedKeys = queries.keys.toList()..sort();
    final Map<String, String> _queries = <String, String>{
      for (final String k in _sortedKeys) encode(k): encode(queries[k]!),
    };
    final List<String> _formalizedStringPair = <String>[
      for (final MapEntry<String, String> e in _queries.entries)
        '${e.key}=${e.value}',
    ];
    final String _formalizedString = _formalizedStringPair.join('&');
    final String _signString = '$method'
        '&${encode('/')}'
        '&${encode(_formalizedString)}';
    final String _signature = encode(
      base64Encode(
        Hmac(
          sha1,
          utf8.encode(accessKeySecret + '&'),
        ).convert(utf8.encode(_signString)).bytes,
      ),
    );

    queries['Signature'] = Uri.decodeFull(_signature);
  }

  static String encode(String value) {
    return Uri.encodeComponent(value)
        .replaceAll('+', '%20')
        .replaceAll('*', '%2A')
        .replaceAll('%7E', '~');
  }
}


example/aliyun_ocr_sdk_example.dart
import 'dart:io';
import 'package:aliyun_ocr_sdk/aliyun_ocr_sdk.dart';
import 'package:aliyun_ocr_sdk/src/aliyun_ocr_sdk_base.dart';
import 'package:dio/dio.dart';

void main() async {

  /// 1. Old Method OCR
  var aliYunSdkOld = aliYunOcrClientOld();

  /// 1.1 Input Your appcode
  aliYunSdkOld.appCode = "YourappCode";
  aliYunSdkOld.generateHeaders();

  print(aliYunSdkOld.headers);

  /// 1.2 Input ImgUrl
  aliYunSdkOld.imgUrl = "https://img.alicdn.com/tfs/TB1Wo7eXAvoK1RjSZFDXXXY3pXa-2512-3509.jpg";
  aliYunSdkOld.generateBodyWithImg();

  /// 1.2 Or Input Img binary
  final value = File("bin/1.png");
  var bytes =await value.readAsBytes();
  aliYunSdkOld.imgBinary = bytes.buffer.asUint8List();
  aliYunSdkOld.generateBodyWithImg();

  // 1.3 Use the post method
  var dio = Dio();
  Response response;
  try {
    response =
    await dio.post(
      aliYunSdkOld.baseUrl,
      data:aliYunSdkOld.body,
      options: Options(
        headers: aliYunSdkOld.headers,
      )
    );
    print(response.statusCode);
    print(response.data);
  } on DioError catch (e) {
    final response = e.response;
    if (response != null) {
      print(response.data);
      print(response.headers);
      print(response.requestOptions);
    } else {
      /// Something happened in setting up or sending the request that triggered an Error
      print(e.requestOptions);
      print(e.message);
    }
  }


  /// 2. New Method OCR
  var aliYunSdkNew = aliYunOcrClientNew();

  /// 2.1 Input Your accessKey
  aliYunSdkNew.accessKeyId = "YouraccessKeyId";
  aliYunSdkNew.accessKeySecret = "YouraccessKeySecret";

  /// 2.2 Input imgUrl
  aliYunSdkNew.imgUrl =
  "https://img.alicdn.com/tfs/TB1Wo7eXAvoK1RjSZFDXXXY3pXa-2512-3509.jpg";

  aliYunSdkNew.queries['AccessKeyId'] = aliYunSdkNew.accessKeyId;
  aliYunSdkNew.queries['Url'] = aliYunSdkNew.imgUrl;

  /// 2.3 Generate Signature and update the queries
  aliYunSdkNew.generateSignature(
      accessKeySecret: aliYunSdkNew.accessKeySecret,
      queries: aliYunSdkNew.queries,
      method: "POST",
  );

  print(aliYunSdkNew.queries);

  // 2.4 Use the post method
  var dio1 = Dio();
  Response response1;
  try {
    response1 =
    await dio1.post(aliYunSdkNew.baseUrl, queryParameters: aliYunSdkNew.queries);
    print(response1.statusCode);
    print(response1.data);
  } on DioError catch (e) {
    final response = e.response;
    if (response != null) {
      print(response.data);
      print(response.headers);
      print(response.requestOptions);
    } else {
      /// Something happened in setting up or sending the request that triggered an Error
      print(e.requestOptions);
      print(e.message);
    }
  }



}
